internal/config/ociartifact/ociartifact.go

```diff
@@ -24,54 +28,135 @@ func New() Impl {
+	SystemContext *types.SystemContext
 
-	// Cleanup has to be called if the artifact is not used any more.
-	Cleanup func()
+	// EnforceConfigMediaType can be set to enforce a specific manifest config media type.
+	EnforceConfigMediaType string
+
+	// MaxSize is the maximum size of the artifact to be allowed to get pulled.
+	// Will be set to a default of 1MiB if not specified (zero) or below zero.
+	MaxSize int
 }
 
 // defaultImpl is the default implementation for the OCI artifact handling.
 type defaultImpl struct{}
 
-// Pull downloads and mounts the artifact content by using the provided ref.
-func (*defaultImpl) Pull(ctx context.Context, sys *types.SystemContext, ref string) (*Artifact, error) {
-	log.Infof(ctx, "Pulling OCI artifact from ref: %s", ref)
+// defaultMaxArtifactSize is the default maximum artifact size.
+const defaultMaxArtifactSize = 1024 * 1024 // 1 MiB
+
+// Pull downloads the artifact content by using the provided image name and the specified options.
+func (*defaultImpl) Pull(ctx context.Context, img string, opts *PullOptions) (*Artifact, error) {
+	log.Infof(ctx, "Pulling OCI artifact from ref: %s", img)
 
-	storeOpts, err := storage.DefaultStoreOptions(false, 0)
+	name, err := reference.ParseNormalizedNamed(img)
 	if err != nil {
-		return nil, fmt.Errorf("get default storage options: %w", err)
+		return nil, fmt.Errorf("parse image name: %w", err)
 	}
 
-	store, err := storage.GetStore(storeOpts)
+	ref, err := docker.NewReference(name)
 	if err != nil {
-		return nil, fmt.Errorf("get container storage: %w", err)
+		return nil, fmt.Errorf("create docker reference: %w", err)
 	}
 
-	runtime, err := libimage.RuntimeFromStore(store, &libimage.RuntimeOptions{SystemContext: sys})
+	src, err := ref.NewImageSource(ctx, opts.SystemContext)
 	if err != nil {
-		return nil, fmt.Errorf("create libimage runtime: %w", err)
+		return nil, fmt.Errorf("build image source: %w", err)
 	}
 
-	images, err := runtime.Pull(ctx, ref, config.PullPolicyAlways, &libimage.PullOptions{})
+	manifestBytes, mimeType, err := src.GetManifest(ctx, nil)
 	if err != nil {
-		return nil, fmt.Errorf("pull OCI artifact: %w", err)
+		return nil, fmt.Errorf("get manifest: %w", err)
+	}
+
+	parsedManifest, err := manifest.FromBlob(manifestBytes, mimeType)
+	if err != nil {
+		return nil, fmt.Errorf("parse manifest: %w", err)
+	}
+	if opts.EnforceConfigMediaType != "" && parsedManifest.ConfigInfo().MediaType != opts.EnforceConfigMediaType {
+		return nil, fmt.Errorf(
+			"wrong config media type %q, requires %q",
+			parsedManifest.ConfigInfo().MediaType, opts.EnforceConfigMediaType,
+		)
+	}
+
+	layers := parsedManifest.LayerInfos()
+	if len(layers) < 1 {
+		return nil, errors.New("artifact needs at least one layer")
+	}
+
+	// Just supporting one layer here. This can be later enhanced by extending
+	// the PullOptions.
+	layer := layers[0]
+
+	bic := blobinfocache.DefaultCache(opts.SystemContext)
+	rc, _, err := src.GetBlob(ctx, layer.BlobInfo, bic)
```

Comment:
If you want, you could still leverage the returned blob size here to short-circuit size check and return early.

Reply:
Yeah I was considering that. But the size only reads the field in the manifest, which can be anything or `-1` (unknown). I'd like to stick to the limit reader to limit the actual size.

Reply:
I see the size being taken from the HTTP response, per:

```go
func getBlobSize(resp *http.Response) int64 {
	size, err := strconv.ParseInt(resp.Header.Get("Content-Length"), 10, 64)
	if err != nil {
		size = -1
	}
	return size
}
```

The call chain is `GetBlob()` -> `getExternalBlob()` -> `getBlobSize()`.

I was thinking something along the lines of:

```go
rc, size, err := src.GetBlob(ctx, layer.BlobInfo, bic)
if err != nil {
	return nil, fmt.Errorf("get layer blob: %w", err)
}
defer rc.Close()

maxArtifactSize := defaultMaxArtifactSize
if opts.MaxSize > 0 {
	maxArtifactSize = opts.MaxSize
}
if size != -1 && size > maxArtifactSize + 1 {
	return nil, fmt.Errorf("exceeded maximum allowed size of %d bytes", maxArtifactSize)
}
```

Basically, the reader passed is reading the content of the HTTP response. So, we might as well leverage the size of the response body (content) if set to avoid having to read it in the first place.

Am I thinking wrong about this?

Reply:
Ah, that's interesting. Yeah I'll add the code path to early exit :+1:

Reply:
Well, if the input is OCI, the expected size is already in the manifest. No need to even trigger a HTTP request.

(It’s still a good idea to use a `LimitReader` because the registry may be malicious. Actually the expected limit could be `min(BlobInfo.Size, maxArtifactSize)`, _typically_ that’s much less than 1 MB.)